File: Data Structures/LinkedList/C#/LinkedList.cs
using System;

namespace DataStructures.Classes
{
    /// <summary>
    /// This class is the LinkedList class
    /// Where each node is a link in the list
    /// Each node has the data and a pointer
    /// to the next node in the list
    /// TODO: implment a iterator interface
    /// </summary>
    /// <typeparam name="T">Generic Type</typeparam>
    public class LinkedList<T>
    {
        private int _size;        // number of items in list
        private Node<T> _head;    // head pointer

        /// <summary>
        /// Default constructor
        /// </summary>
        public LinkedList()
        {
            _size = 0;
            _head = null;
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="data">The value to init the list with</param>
        public LinkedList(T data)
        {
            // set the size
            _size = 1;
            // set the head to this new node
            _head = new Node<T>(data, null);
        }

        /// <summary>
        /// Gets the number of items in the linked list
        /// </summary>
        /// <returns>Size of linked list</returns>
        public int Size()
        {
            return _size;
        }

        /// <summary>
        /// Checks if the linked list is empty
        /// </summary>
        /// <returns>true if empty</returns>
        public bool IsEmpty()
        {
            return _size == 0;
        }

        /// <summary>
        /// Gets the value from the front of the linked list
        /// </summary>
        /// <returns>Value from the front of the linked list</returns>
        public T Front()
        {
            return _head.GetData();
        }

        /// <summary>
        /// Pushes a new node to the front of the linked list
        /// </summary>
        /// <param name="data">The value to add to the linked list</param>
        public void PushFront(T data)
        {
            // increment the size
            _size++;
            
            // create a new node to push to the front
            var current = new Node<T>(data, null);
            // set the current to head
            current.SetNext(_head);
            // set the head to current
            _head = current;
        }

        /// <summary>
        /// Pushes a new node to the end of the linked list
        /// </summary>
        /// <param name="data">The new node data to put at end of linked list</param>
        public void PushBack(T data)
        {
            // init a pointer to the head
            var current = _head;

            // loop as long as next node isnt null
            while (current.GetNext() != null)
                // move to next
                current = current.GetNext();
            
            // set the next node with data and the next's next node as null
            current.SetNext(new Node<T>(data, null));
        }

        /// <summary>
        /// Gets the value from the back of the list
        /// </summary>
        /// <returns>Value from the back of the linked list</returns>
        /// <exception cref="ArgumentOutOfRangeException">Throws exception if list is empty</exception>
        public T Back()
        {
            // if head is null, list is empty
            if (_head == null)
                throw new ArgumentOutOfRangeException($"Linked List is empty!");
            // init a pointer to head
            var current = _head;

            // loop while next node isn't null
            while (current.GetNext() != null)
            {
                // traverse to next node
                current = current.GetNext();
            }

            // return the value from the end of the list
            return current.GetData();
        }

        /// <summary>
        /// Gets the value from the front of list and removes from list
        /// </summary>
        /// <returns>Value from the front of the Linked list</returns>
        /// <exception cref="ArgumentOutOfRangeException">Throws exception if linked list is empty</exception>
        public T PopFront()
        {
            // if head is null, empty list
            if (_head == null)
                throw new ArgumentOutOfRangeException($"Linked List is empty!");

            // store the data from the head
            var data = _head.GetData();
            // point the head to the next node
            _head = _head.GetNext();
            // decrement the size
            _size--;
            // return the data
            return data;
        }

        /// <summary>
        /// Gets the value from the end of list and returns it
        /// </summary>
        /// <returns>Value from end of linked list</returns>
        /// <exception cref="ArgumentOutOfRangeException">Throws exception if list is empty</exception>
        public T PopBack()
        {
            // if the head is null, throw exception
            if (_head == null)
                throw new ArgumentOutOfRangeException($"Linked List is empty!");

            // init a node to hold previous node
            var prev = new Node<T>();
            
            // cache head
            var current = _head;

            // while next node isn't null
            while (current.GetNext() != null)
            {
                // store prev
                prev = current;
                // move current to next
                current = current.GetNext();
            }

            // get the data from the current
            var data = current.GetData();
            // set the prev's next node to null
            prev.SetNext(null);
            // decrement the size of list
            _size--;
            // return the data
            return data;
        }

        /// <summary>
        /// Gets the value from the node of the index passed
        /// </summary>
        /// <param name="index">Location of node</param>
        /// <returns>Value at the index supplied</returns>
        /// <exception cref="ArgumentOutOfRangeException">Throws exception if linked list is empty</exception>
        public T ValueAt(int index)
        {
            // if head is null, empty list
            if (_head == null)
                throw new ArgumentOutOfRangeException($"Linked List is empty!");

            // boolean if found in list
            var found = false;

            // store head for traversal
            var current = _head;
            // loop over linked list
            for (int i = 0; current.GetNext() != null; i++)
            {
                // if i is index, break out we found
                if (i == index)
                {
                    found = true;
                    break;
                }

                // move current to next
                current = current.GetNext();
            }
            // if the index wasn't in list
            if (!found)
                throw new ArgumentOutOfRangeException($"Index is out of range of the linked list");
            // return the data at current
            return current.GetData();
        }

        /// <summary>
        /// Removes the node with the value specified
        /// does nothing if value is not in linked list
        /// </summary>
        /// <param name="data">Value to remove</param>
        /// <exception cref="ArgumentOutOfRangeException">Throws exception if empty</exception>
        public void Remove(T data)
        {
            // if head is null, list is empty
            if (_head == null)
                throw new ArgumentOutOfRangeException($"Linked List is emtpy!");

            // pointers to traverse list
            var current = _head;
            var prev = current;

            // loop through list
            while (current != null)
            {
                // if current has value to remove, break
                if (current.GetData().Equals(data))
                    break;

                // set prev to the current
                prev = current;
                // move current to the next
                current = current.GetNext();
            }
            // set the prev node's next to current's next
            // effectively removing the current node
            prev.SetNext(current?.GetNext());
        }

        /// <summary>
        /// Reverse the Linked List
        /// </summary>
        /// <exception cref="ArgumentOutOfRangeException">Throws exception if Linked List is empyt</exception>
        public void Reverse()
        {
            // if the head is null, list is empty
            if (_head == null)
                throw new ArgumentOutOfRangeException($"List is empty!");

            // create a pointer to head
            var current = _head;
            // declare a node to hold the next
            Node<T> next;
            // init a prev node to null
            Node<T> prev = null;

            // while current isn't null
            while (current != null)
            {
                // set the next node to current's next
                next = current.GetNext();
                // set current's next to prev node
                current.SetNext(prev);
                // point prev to current
                prev = current;
                // point current to next
                current = next;
            }

            // set the head to prev
            _head = prev;
        }
        
        /// <summary>
        /// The To String
        /// </summary>
        /// <returns>The string of the list</returns>
        public override string ToString()
        {
            // init a string
            var output = "";
            // init a pointer to the head
            var current = _head;
            // loop while next isn't null
            while (current.GetNext() != null)
            {
                // append the data and an arrow denoting pointing to next node
                output += current.GetData() + "->";
                // move to next
                current = current.GetNext();
            }

            // append last data to string
            output += current.GetData();
            // return the string
            return output;
        }

        #region Node

        /// <summary>
        /// This is the node class that represents
        /// each node of the linked list
        /// </summary>
        /// <typeparam name="TS">Generic Type</typeparam>
        private class Node<TS>
        {
            private TS _data;
            private Node<TS> _next;
        
            public Node()
            {
                _next = null;
            }

            public Node(TS data, Node<TS> node)
            {
                _data = data;
                _next = node;
            }

            public void SetNext(Node<TS> node)
            {
                _next = node;
            }

            public void SetData(TS data)
            {
                _data = data;
            }

            public Node<TS> GetNext()
            {
                return _next;
            }

            public TS GetData()
            {
                return _data;
            }

        }

        #endregion

    }

    
}

File: Data Structures/LinkedList/C/linkedList.c
#include "linkedList.h"

// This method returns the length of the list
// by traversing the list
int length(struct Node *head) {
    // create a current node to traverse
    struct Node *current = head;
    // declare a counter
    int count;
    count = 0;
    // loop through list as long as current
    // is not NULL (which would be the end)
    while (current != NULL) {
        // increment the counter
        count++;
        // set current to the next
        current = current->next;
    }
    // return the count
    return count;
}

// This method loops through the list
// and prints out each datum at each node
void printList(struct Node *head) {
    // get the first node
    struct Node *current = head;
    // if the head is pointing to NULL
    if (head == NULL) {
        // output info and return
        printf("Linked List is empty\n");
        return;
    }
    // init a counter
    int node = 0;
    // while the current is not null
    while (current != NULL) {
        // output the current node and it's data
        printf("Node %i : %i\n", node, current->data);
        // move to next node
        current = current->next;
        // increment counter
        node++;
    }
}

// This method adds a new node to the first
// position.  Points head to this new node
// and points this new node's next to head's next
void pushFront(struct Node **head, int data) {
    // init a new node, and a pointer
    struct Node *newNode, *current;
    newNode = (struct Node *)malloc(sizeof(struct Node));
    // if new node is null
    if (!newNode) {
        // output error, memory
        error("pushFront");
        // return from method
        return;
    }
    // set the data on the new node
    newNode->data = data;
    // set the current node to head
    current = *head;
    // set new node's next to current (what head is pointing to)
    newNode->next = current;
    // point head to the new node
    *head = newNode;
} 

void pushBack(struct Node **head, int data) {

    // init new node and init pointers
    struct Node *newNode, *current, *prev;
    newNode = (struct Node *) malloc(sizeof(struct Node));

    // if new node is null, memory error and return
    if (!newNode) {
        error("pushBack");
        return;
    }

    // set p to head
    current = *head;
    // if the head is pointing at null
    // empty list, so call pushFron
    if (current == NULL) {
        pushFront(head, data);
        return;
    }
    // set the new node's data
    newNode->data = data;

    // traverse the list
    while (current != NULL) {
        // save current node
        prev = current;
        // set p to next
        current = current->next;
    }
    // set the prev node's next to the new node
    prev->next = newNode;
    // set the new node's next to current (should be null for tail)
    newNode->next = current;
}

// This method will insert a new node at the location
// specified as param loc, 0 based
void insertAt(struct Node **head, int loc, int data) {

    // initialize new node, declare pointers
    struct Node *newNode, *current, *prev;
    newNode = (struct Node *) malloc(sizeof(struct Node));

    // if new node is null, memory error
    if (!newNode) {
        // call helper method for debugging memory issues
        error("insertAt");
        // return from this method
        return;
    }

    // set the current to the head
    current = *head;

    // if user set loc as 0 or current (head) is null
    if (loc == 0 || current != NULL) {
        // add the node to the front
        pushFront(head, data);
        return;
    }

    // set the new node's data
    newNode->data = data;
    // init a pos counter
    int pos = 0;

    // while current is not null, and the pos is 
    // less than the location
    while (current != NULL && pos < loc) {
        // increment position
        pos++;
        // set the prev
        prev = current;
        // set current to next
        current = current->next;
    }
    // if the location is greater than how many nodes
    // this linked list has, this above while loop
    // will fall of, because checking current to be not null
    // if current is null, we are at the end of the list
    // and below it will properly set the node at the end

    // finally set the prev's next to the new node
    prev->next = newNode;
    // set the new node's next to the current
    newNode->next = current;
}

// This method removes the first node in
// the list, and returns the value
int popFront(struct Node **head) {
    // if head is pointing to nothing
    if ((*head) == NULL) {
        // output that list is empty
        printf("List is empty\n");
        // return a flag that list is empty
        return -1;
    } 
    // get the data from the head (dereference pointer to pointer)
    int data = (*head)->data;
    // set head to to the next, effectively next node is not accessable
    *head = (*head)->next;
    // return the data
    return data;
}
// This method removes the last node
// in the list, and returns that datum
int popBack(struct Node **head) {
    // if the head is pointing to null
    if ((*head) == NULL) {
        // output that list is empty
        printf("List is empty\n");
        // return a flag
        return -1;
    }
    // declare pointers for traversal
    struct Node *prev, *current;
    // set the current pointer to head
    current = *head;
    // declare var to hold the datum
    int data;
    // while the next is not null
    while (current->next != NULL) {
        // hold prev node
        prev = current;
        // set current to next
        current = current->next;
        // get the datum from last node
        data = current->data;
    }
    // set the previous node's next to NULL
    prev->next = NULL;
    // free up the memory of the current (which was last)
    free(current);
    // return the data
    return data;
}

// This method removes a specific node of the key specified
void removeNode(struct Node **head, int key) {
    // init an int(bool) if found
    short found = 0;
    // declare pointers for the previous and current
    struct Node *prev, *current;
    // set the current to the head
    current = *head;

    // while the next node is not the end
    while (current->next != NULL) {
        // if the current is the key
        if (current->data == key) {
            // set flag that found
            found = 1;
            // break from loop
            break;
        }
        // save current to prev
        prev = current;
        // advance to next node
        current = current->next;
    }
    // if key not in list
    if (!found) {
        // output error message
        printf("Node with data of %i is not in the list\n", key);
        // return from method
        return;
    }
    // set the next node, skipping node to delete
    prev->next = current->next;
    // free the memory of current
    free(current);
}

// This method deletes the entire linked list
void deleteList(struct Node **head) {
    // declare pointers to traverse list
    struct Node *iter, *aux;
    // set the iterator to the head
    iter = *head;
    // while the iterator is not pointing to NULL
    while (iter != NULL) {
        // set the auxillary node to the iter's next
        aux = iter->next;
        // free the iterator from memory
        free (iter);
        // set the auxillary to the iterator
        iter = aux;
    }
    // finally set the head to NULL
    *head = NULL;
}

// This method reverses the linked list
void reverse(struct Node **head) {
    // declare pointers for traversal/reversal
    struct Node *prev, *current, *next;
    // point current node to the head
    current = *head;
    // while current node is not null
    while (current) {
        // point the next to current's next
        next = current->next;
        // set the current next as prev (first time will be null)
        current->next = prev;
        // point prev to current
        prev = current;
        // point current to next
        current = next;
    }
    // finally point the head to previous
    *head = prev;
}

// helper method that prints out error memory
// param: string of method for debugging purposes
void error(char *method) {
    fprintf(stderr, "%s. Memory error: cannot initlize new node\n", method);
}
